File: scripts/query_api_show_availability.py
import polars as pl
from datetime import datetime, date
from google.cloud import storage
from functions.show_availability_api.utils import *

def save_to_google_storage(dat, bucket):
    # First save locally in parquet format
    """
    Save a Polars DataFrame to Google Cloud Storage, with the current date included in the file name.

    Inputs:
    - dat: polars.DataFrame - The dataframe to be saved
    - bucket: str - The Google Cloud Storage bucket to save to

    Returns: None
    """
    
    file_name = f"show-availability-query-date-{str(date.today())}.parquet"
    path_to_file = f"data/{file_name}"
    dat.write_parquet(path_to_file)

    # Then upload to Google Cloud Storage from local
    try:
        storage_client = storage.Client()
        bucket = storage_client.bucket(bucket)
        blob = bucket.blob(file_name)
        blob.upload_from_filename(path_to_file)
        print(f"Saved {file_name} to Google Cloud Storage")
    except Exception as e:
        print(f"Failed to save {file_name} to Google Cloud Storage")
        print(e)

# ==============================================================================

if __name__ == "__main__":
    # Load secrets from CSV file
    secrets = pl.read_csv("secrets.csv")
    affiliate_id = secrets.filter(pl.col("secretName") == "affiliateId").get_column("secretValue")[0]
    
    # We shall query the API for the following list of shows
    show_list = pl.DataFrame({
        "show_name": ["A Christmas Carol(ish)", "Oedipus (by Robert Icke)"],
        "show_id": [42462, 41707],
        "date_start": ["20241101", "20241101"],
        "date_end": ["20241231", "20241231"]
    })

    # Get API responses for the list of shows
    api_responses = [query_api_for_show_availability(row["show_id"], row["date_start"], row["date_end"], affiliate_id) for row in show_list.iter_rows(named=True)]

    # Parse API responses into Polars DataFrames
    parsed_responses = [parse_api_response(api_response) for api_response in api_responses if api_response is not None]

    # If not empty, save the parsed results into a single dataframe
    if parsed_responses == []:
        print("No API response parsed. Nothing to save.")
    else: 
        # Concatenate into a single DataFrame
        dat_parsed = pl.concat(parsed_responses, how="vertical")
        
        print("Head of the resulting DataFrame:")
        print(dat_parsed.glimpse())

        # Save the dataframe in parquet format locally, and also upload to google cloud storage
        save_to_google_storage(dat_parsed, "raw-todaytix-api-show-availability")

File: scripts/query_api_show_availability.R
library(here)
library(httr2)
library(glue)
library(tibble)
library(purrr)
library(dplyr)
library(nanoparquet)

## == SETUP / LOAD SECRETS =====================
dat_secrets <- read.csv(here("secrets.csv"))
AFFILIATE_ID <- dat_secrets[dat_secrets$secretName == "affiliateId"]$secretValue


## == DEFINE FUNCTIONS =========================

#' Query the API for show availability information
#' @param show_id The ID of the show
#' @param date_start The start date of the availability query
#' @param date_end The end date of the availability query
#' @param affiliate_id The client AffiliateID assigned by TTG
#' @return REST API response, or NULL if error
query_api_for_show_availability <- function(show_id, date_start, date_end, affiliate_id=AFFILIATE_ID) {
  # Form the query URL
  URL_AVAILABILITY <- glue("https://inventory-service.tixuk.io/api/v4/availability/products/{show_id}/quantity/1/from/{date_start}/to/{date_end}/detailed")

  # Build the API request, including headers
  api_request <- request(URL_AVAILABILITY) |>
    req_headers(affiliateId = AFFILIATE_ID)

  # Perform the API request
  tryCatch(res <- req_perform(api_request),
    error = \(e) {print(glue("Failed to query the API for show ID {show_id}")); print(e); res <- NULL})
}


#' Parse the API response into a tibble. The function currently expects the body of the API response
#' contains partiicular fields. It will break if that changes.
#' @param api_response The REST API response
#' @return A tibble
parse_api_response <- function(api_response) {

  if (!is.null(api_response)) {
        
    # Convert body of the response to JSON
    response_content_json <- resp_body_json(api_response)

    # Extract show availability information
    response_show_availability <- response_content_json$response$availability

    # ... and then use purrr:map to convert into a tibble
    dat_show_availability <- tibble(
      performanceTime = map_chr(response_show_availability, "datetime", .default=NA_character_),
      performanceType = map_chr(response_show_availability, "performanceType", .default=NA_character_),
      availableSeatCount = map_int(response_show_availability, "availableSeatCount", .default=NA_integer_),
      largestLumpOfTickets = map_int(response_show_availability, "largestLumpOfTickets", .default=NA_integer_),
      minPrice = map_int(response_show_availability, "minPrice", .default=NA_integer_),
      maxPrice = map_int(response_show_availability, "maxPrice", .default=NA_integer_),
      discountAvailable = map_lgl(response_show_availability, "discount", .default=NA),
      currency = map_chr(response_show_availability, "currency", .default=NA_character_)
    )

    # Also extract the requested show ID as recorded in the response
    response_show_id <- response_content_json$response$show
    # Also extract the request timestamp (as recorded in the response)
    response_timestamp <- as.POSIXct(resp_date(api_response))

    # Finally, form the final tibble
    dat_show_availability <- dat_show_availability |>
      mutate(
        performanceTime = as.POSIXct(performanceTime, tz="UTC", format="%Y-%m-%dT%H:%M:%S"),
        requestTime = response_timestamp,
        showId = response_show_id
    ) 
  } else {
    res <- NULL
  }
}


## == Query the API for a number of shows =============================
# A list of shows
show_list <- tribble(
  ~show_name, ~show_id, ~date_start, ~date_end,
  "A Christmas Carol(ish)", "42462", "20241101", "20241231",
  # "Test for error", "111", "20241101", "20241231",
  "Oedipus (by Robert Icke)", "41707", "20241101", "20241231"
)

# Query the API
api_response <- pmap(
  list(show_list$show_id, show_list$date_start, show_list$date_end),
  query_api_for_show_availability
)

# Parse the API response for all shows, and assemble into a single tibble
dat_parsed <- map(api_response, parse_api_response) |>
  list_rbind() |>
  as_tibble()

# Save on disk as parquet

query_execution_date <- max(dat_parsed$requestTime) |> as.Date()

write_parquet(dat_parsed,
  glue(here("data", "show-availability-query-date-{query_execution_date}.parquet")))


# ===================================
# # This is how one should be able to read all the files back into a tibble
files_to_read <- list.files(here("data/"), pattern="*.parquet", full.names=TRUE)
dat_read <- map(files_to_read, ~read_parquet(.)) |>
  list_rbind()|> 
  as_tibble()

